### PR TITLE
[SPARK-16590][SQL] Improve LogicalPlanToSQLSuite to check generated SQL directly

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

This PR improves `LogicalPlanToSQLSuite` to check the generated SQL directly by **structure**. So far, `LogicalPlanToSQLSuite` relies on  `checkHiveQl` to ensure the **successful SQL generation** and **answer equality**. However, it does not guarantee the generated SQL is the same or will not be changed unnoticeably.
## How was this patch tested?

Pass the Jenkins. This is only a testsuite change.

### CODE REVIEW

**Comment by rxin** [comment]
hm the problem with this approach is that we'd need to spend a lot of time to update test cases whenever we change sql generation slightly. I think in order to do this, we should put the generated sql in  files and then have a way to regenerate all the sql queries in bulk.

**Comment by dongjoon-hyun** [comment]
Oh, thank you for fast review!
Yep. I will update like that.
The purpose of this PR is having stronger `LogicalPlanToSQLSuite`.

**Comment by gatorsmile** [comment]
It sounds like the latest changes are unable to resolve Reynold's concern: how to regenerate all the expected SQL queries in bulk? 

You know, for native view support, we are not generating optimal SQL statements. You can see the generated SQL is verbose and not readable. However, these SQL statements can be optimized by Catalyst optimizer at runtime. Thus, I have the same concern. This approach is not maintainable

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala [hunk 1]
```diff
@@ -76,7 +79,29 @@ class LogicalPlanToSQLSuite extends SQLBuilderTest with SQLTestUtils {
     }
   }
 
-  private def checkHiveQl(hiveQl: String): Unit = {
+  // Used for generating new query answer files by saving
+  private val saveQuery = false
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: Hi, @gatorsmile .
This is my answer to that. :)

**Comment by dongjoon-hyun** [comment]
First of all, you can update the whole query set by one flag. So, maintainability is no more difficult issue.

I made this PR because `SQL generation` is currently fragile as [you said yesterday](https://github.com/apache/spark/pull/14132#issuecomment-233107976).

We need to prevent unintentional and accidental changes on that before both Hint or [SPARK-16576](https://issues.apache.org/jira/browse/SPARK-16576).

IMO, this is a **correctness** issue we should resolve. I hope this PR protects Spark from **me**. :)

**Comment by gatorsmile** [comment]
Comparing the SQL statement strings is horrible to me. 

I have a different approach to verify the **correctness**. How about compare the optimized plans, which can tolerate more slight changes that do not affect the correctness?

**Comment by dongjoon-hyun** [comment]
:) Sorry, but I don't think so.
* At every level, we need to prove correctness. The tolerance you want is the result of unpredictable removal of Optimizer.
* Also, this is `LogicalPlanToSQLSuite` . SQL statement comparison is not a good way, but the only correct way to this module.

**Comment by gatorsmile** [comment]
I can understand the advantage of comparing SQL statement strings, but the cost is higher than the benefit especially for a small group of key developers who are maintaining the Spark code everyday. 

In some commercial RDBMS, they have a whole team for developing and maintaining SQL generation. However, I am not sure if the Spark community can afford it. You know, SQL generation is only used for native view support so far.

**Comment by dongjoon-hyun** [comment]
I'm not sure what cost exactly do you mean.
It's a flag, isn't it? Also, if someone change SQLBuilder, it should be tested by the generated SQLs, not the query execution result.
For example, BROADCAST hints will be tested here as comments after merging this PR.

**Comment by gatorsmile** [comment]
`BROADCAST hints` is special. We need to check the statements.

My concern is we might spend a lot of times when making changes that impact the SQL statement strings but it does not change the correctness of SQL builder.

**Comment by dongjoon-hyun** [comment]
First, we will make more HINTS. I made a general HINT syntax for that. We are opening the door together.

Second, when you add your new testcase here, the additional cost is just **one second** to generate the file.

Third, when you develops your PR, this will save much time. I'm already using this to save my time. At this time, we need to verify correctness at every level. If you use optimized result or the execution result, that is a naive test. You know that.

**Comment by dongjoon-hyun** [comment]
If I miss some cost you concern, could you give some examples? I'm still not sure what you mean by cost.

**Comment by rxin** [comment]
@dongjoon-hyun can you update the classdoc for the test suite to explain how to generate the expected results?

**Comment by rxin** [comment]
Also personally I find it difficult to look at the result without looking at the original SQL. Can we update the result file to show the following format?

```
original query
---------------------
generated query
```

**Comment by dongjoon-hyun** [comment]
Oh, sure. I'll update the classdoc and change the file format.

**Comment by rxin** [comment]
Checking the generated SQL is a good thing to do - we just need to make sure it is maintainable (i.e. have tooling to automatically regenerate results).

Once we are done with this, we should also do similar things for the parser -- I think the unit tests for parsers don't currently check for the plan, but only checks whether the parser can parse it.

**Comment by gatorsmile** [comment]
For example, if we change generation of alias names, most of test cases will fail.

**Comment by gatorsmile** [comment]
Yeah, we need to do it for `Parser`. That part is very independent. The changes on the other parts will not affect it.

**Comment by gatorsmile** [comment]
Regenerating SQL statements can simplify the work, but the reviewers have to be careful when reviewing the statement changes. Those statements could be very complex and hard to read.

**Comment by rxin** [comment]
Since the purpose of the function here is to generate SQL, I'd say it's important to actually show the generated SQL in reviews.

That said, if we can also get something even more automated (e.g. asserting that the optimized plan equals) in addition to this, then it'd be even more robust!

**Comment by dongjoon-hyun** [comment]
Now, the answer files have both original and generated queries and the description of regenerating answer sets is added to `LogicalPlanToSQLSuite` classdoc.

File: sql/hive/src/test/resources/sqlgen/agg1.sql [hunk 2]
```diff
@@ -0,0 +1,3 @@
+SELECT COUNT(value) FROM parquet_t1 GROUP BY key HAVING MAX(key) > 0
+--------------------------------------------------------------------------------
+SELECT `gen_attr` AS `count(value)` FROM (SELECT `gen_attr` FROM (SELECT count(`gen_attr`) AS `gen_attr`, max(`gen_attr`) AS `gen_attr` FROM (SELECT `key` AS `gen_attr`, `value` AS `gen_attr` FROM `default`.`parquet_t1`) AS gen_subquery_0 GROUP BY `gen_attr` HAVING (`gen_attr` > CAST(0 AS BIGINT))) AS gen_subquery_1) AS gen_subquery_2
```
Comment (rxin) [inline sql/hive/src/test/resources/sqlgen/agg1.sql]: can you add a new line to the end? this little red thing on github is annoying

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala [hunk 3]
```diff
@@ -76,7 +85,34 @@ class LogicalPlanToSQLSuite extends SQLBuilderTest with SQLTestUtils {
...
+        val sqls = answerText.split(separator)
+        if (sqls.length == 2) {
+          val normalizedExpectSQL = sqls(1).trim()
+          assert(normalizedGenSQL == normalizedExpectSQL)
+        }
+      }
+    }
+  }
+
+  private def checkHiveQl(hiveQl: String, answerFile: String = null): Unit = {
```
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: while you at it, rename checkHiveQl to something else (e.g. checkSQL).

Also why should answerFile be null?

**Comment by rxin** [comment]
@dongjoon-hyun can you take a look at @gatorsmile's suggestion to check optimized logical plan and see if it is feasible?

File: sql/hive/src/test/resources/sqlgen/agg1.sql (hunk 2 above)
Comment (dongjoon-hyun) [inline sql/hive/src/test/resources/sqlgen/agg1.sql]: Oh, I see. Sure.

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala (hunk 3 above)
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: Oh, it's due to that I found that some Windows queries generated SQL is non-deterministic. The order in the windows functions varies sometime. As a result, the test sometime passes and sometime not. Until now, those testcases are not transformed into this style in this PR.
I'm planing to investigate the root cause later.

**Comment by dongjoon-hyun** [comment]
It's definitely a solution. In fact, I tried that first before making this PR, of course, not completely, but as a feasibility tests. It has its own pros and cons, too.

For example, if we add new optimizers like `OptimizeIn` by removing duplications, we should change here too. (If there is a case here.) This is the same situation because we didn't check the any plans or queries until now. We only check the parseability and the result.

I'm sure that all attempts beyond the existing suite have the same side-effects.

This PR aims to ensure the correctness before making the changes on SQLBuilder and Hint, or other refactoring. It has its meaning since the current testsuite is fragile as we know.

You may want to change the level of importance or test coverage for any convenience. I understand the intention of @gatorsmile , too. 

But, IMHO, it's difficult to refactor some module without a proper testsuite. At the worst case, your can remove the second parameters easily when we reach more stable level. It's similar with changing `test()` into `ignore()`.

Lastly, let me suggest like this. My opinion is having this now for further refactoring. After all SQL functions moves into each LogicalPlans unittests, we can downgrade this to compare optimized plan in most cases (not all cases).

How do you think about this way, @rxin and @gatorsmile ?

**Comment by rxin** [comment]
@dongjoon-hyun i'm not against merging this first. Just want to see if we can improve it further.

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala [hunk 4]
```diff
@@ -17,12 +17,21 @@
...
+import java.nio.file.{Files, Paths}
+
 import scala.util.control.NonFatal
 
 import org.apache.spark.sql.Column
 import org.apache.spark.sql.functions._
 import org.apache.spark.sql.test.SQLTestUtils
 
+/**
+ * A testsuite for LogicalPlan-to-SQL conversion.
```
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: testsuite -> test suite

Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: also add something like

"Each query has a golden generated SQL file in test/resources/sqlgen. The test suite also has built-in functionality to automatically generate these golden files. To do so, set the saveQuery flag to true and
run the test suite. After generating the golden files, make sure to set the flag to false."
```

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala [hunk 5]
```diff
@@ -76,22 +85,51 @@ class LogicalPlanToSQLSuite extends SQLBuilderTest with SQLTestUtils {
...
+  // Used for generating new query answer files by saving
+  private val saveQuery = false
+
+  /**
+   * Compare the generated SQL with the expected answer string.
+   * Note that there exists a normalization for both arguments for the convenience.
+   * - Remove the id from the generated attributes, e.g., `gen_attr_1` -> `gen_attr`.
+   */
+  private def checkSQLStructure(originalSQL: String, convertedSQL: String, answerFile: String) = {
+    val normalizedGenSQL = convertedSQL.replaceAll("`gen_attr_\\d+`", "`gen_attr`")
```
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: I'd say we should not include this.

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala [hunk 6]
```diff
@@ -76,22 +85,51 @@ class LogicalPlanToSQLSuite extends SQLBuilderTest with SQLTestUtils {
...
+    if (answerFile != null) {
+      val path = s"src/test/resources/sqlgen/$answerFile.sql"
```
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: use classloader getResource to get the file, rather than hardcoding the path.

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala (hunk 4 above)
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: Sure.

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala (hunk 5 above)
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: Sorry, what do you mean? The omitting number from gen_attr_123?

File: sql/hive/src/test/resources/sqlgen/agg1.sql [hunk 7]
```diff
@@ -0,0 +1,3 @@
+SELECT COUNT(value) FROM parquet_t1 GROUP BY key HAVING MAX(key) > 0
```
Comment (rxin) [inline sql/hive/src/test/resources/sqlgen/agg1.sql]: one more thing -- can the generated file actually say

```
-- This file is automatically generated by LogicalPlanToSQLSuite
query
--------------------------------------------------------------------------------
generated query
```

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala (hunk 6 above)
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: No problem.

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala [hunk 8]
```diff
@@ -76,22 +85,51 @@ class LogicalPlanToSQLSuite extends SQLBuilderTest with SQLTestUtils {
...
+        val sqls = answerText.split(separator)
+        if (sqls.length == 2) {
+          val normalizedExpectSQL = sqls(1).trim()
+          assert(normalizedGenSQL == normalizedExpectSQL)
+        }
+      }
+    }
+  }
+
+  private def checkSQL(sqlString: String, answerFile: String = null): Unit = {
```
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: When is answerFile null? You should document this. Also why does it default to null?

Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: It'd be great to document what this function does too, e.g. 

1. Checks generated SQL against golden files.
2. Verifies the execution result stays the same.

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala [hunk 9]
```diff
@@ -102,7 +140,7 @@ class LogicalPlanToSQLSuite extends SQLBuilderTest with SQLTestUtils {
            |$convertedSQL
            |
            |# Original HiveQL query string:
```
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: HiveQL -> SQL

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala [hunk 10]
```diff
@@ -76,22 +85,51 @@ class LogicalPlanToSQLSuite extends SQLBuilderTest with SQLTestUtils {
...
+
+  private def checkSQL(sqlString: String, answerFile: String = null): Unit = {
+    val df = sql(sqlString)
 
     val convertedSQL = try new SQLBuilder(df).toSQL catch {
       case NonFatal(e) =>
         fail(
           s"""Cannot convert the following HiveQL query plan back to SQL query string:
              |
              |# Original HiveQL query string:
```
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: Hive -> SQL

Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: Can you find other instances of HiveQL and replace it with SQL.

**Comment by rxin** [comment]
@dongjoon-hyun please also implement unit tests for the checking functionality. In particular, we want to make sure the test actually fails if the generated SQL does not match the golden file.

**Comment by rxin** [comment]
@dongjoon-hyun what problem did you run into when you tried comparing optimized plans?

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala (hunk 5 above)
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: are we not using stable identifiers for those? How come we have this normalization process?

File: sql/hive/src/test/resources/sqlgen/agg1.sql (hunk 7 above)
Comment (dongjoon-hyun) [inline sql/hive/src/test/resources/sqlgen/agg1.sql]: It looks much better. Sure.

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala (hunk 9 above)
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: Oh. Yep.

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala (hunk 10 above)
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: Yep.

**Comment by dongjoon-hyun** [comment]
Hmm, I didn't remember correctly, but the result were different. For the Hint test cases, I used SQLConf to setup the testsuite because I knew that the optimizer would work differently.

**Comment by dongjoon-hyun** [comment]
For the self unittest, sure. I did it manually, but it would be better if we had it, too.

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala (hunk 8 above)
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: I added like this. Is it okay for you?
```scala
  /**
   * 1. Checks if SQL parsing succeed.
   * 2. Checks if SQL generation succeed.
   * 3. Checks the generated SQL against golden files.
   * 4. Verifies the execution result stays the same.
   */
```

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala (hunk 5 above)
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: Yes. It seems that only `gen_subquery_XXX` is stable and `gen_attr_YYY` is not, currently.

Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: Actually, as you guess in the function name, it compares the `Structure`.
We remove the id here, but the alias is remaining. So, the output column names are checked correctly for aliases.

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala (hunk 6 above)
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: For this one, it seems we need to use that path when we save.
When `getClass.getClassLoader.getResource` returns `test-classes/sqlgen`.
Is it okay to keep that?

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala (hunk 8 above)
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: That's good.

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala (hunk 6 above)
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: just use a dummy file (e.g. sqlgen/README.md) to get the path using getResource then.

It's probably good to have README.md in sqlgen to explain what that folder is anyway.

**Comment by gatorsmile** [comment]
When the optimized plans are different, we need to check whether the optimized plan of the generated SQL perform worse than the original one. If so, we need to improve the SQL generation logics.

**Comment by dongjoon-hyun** [comment]
The followings are updated.
- Adds a heading line in golden SQL files.
- Adds more documents and replace `HiveQL` with `SQL`. The remaining `hive` are only the one in test input SQL.
- Adds a testcase for `checkSQL`: SQL Parsing, Test File Existence, SQL Generation.
  - Only execution result checking is omitted, because the query structure is already the same.

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala (hunk 5 above)
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: Can we make gen_attr_YYY stable? (can submit a separate pr)

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala [hunk 11]
```diff
@@ -76,22 +88,58 @@ class LogicalPlanToSQLSuite extends SQLBuilderTest with SQLTestUtils {
...
+      val path = s"src/test/resources/sqlgen/$answerFile.sql"
+      val separator = "-" * 80
+      if (saveQuery) {
+        val header = "-- This file is automatically generated by LogicalPlanToSQLSuite."
+        val answerText = s"$header\n${originalSQL.trim()}\n${separator}\n$normalizedGenSQL\n"
+        Files.write(Paths.get(path), answerText.getBytes(StandardCharsets.UTF_8))
+      } else {
+        val answerText = new String(Files.readAllBytes(Paths.get(path)), StandardCharsets.UTF_8)
+        val sqls = answerText.split(separator)
+        if (sqls.length == 2) {
```
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: shouldn't we do an assert here instead?

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala [hunk 12]
```diff
@@ -101,95 +149,125 @@ class LogicalPlanToSQLSuite extends SQLBuilderTest with SQLTestUtils {
...
+           |# Original SQL query string:
+           |$sqlString
            |
            |# Resolved query plan:
            |${df.queryExecution.analyzed.treeString}
          """.stripMargin, cause)
     }
   }
 
+  test("checkSQL") {
```
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: add some comment inline explaining what we are actually testing here - also maybe rename the test name to "test behavior for the golden file test harness"

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala (hunk 5 above)
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: Sure. I think so. It would be great issue.

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala (hunk 12 above)
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: what i meant about inline comment was something like this

```
test("test behavior for the golden file test harness") {
  // 1. Test should fail if the SQL query cannot be parsed.
  val m = intercept[ParseException] {
    ...
  }
  assert(...)

  // 2. Test should fail if the golden file cannot be found
  ...

}
```

Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: Actually maybe these should be 4 test cases instead.

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala (hunk 6 above)
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: Ur, I think any resource files under `resources` are copied into `test-classes` with classes.

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala (hunk 11 above)
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: No problem.

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala (hunk 12 above)
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: Oh, I see. I'll fix like that.

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala (hunk 6 above)
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: What do you mean? What I meant is to create a README.md file in sqlgen directory, and at runtime we can get the location of sqlgen directory by getResource("sqlgen/README.md")

Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: What I mean is `sqlgen` is currently `sql/hive/src/test/resources/sqlgen`. So, `sql/hive/src/test/resources/sqlgen/README.md` is also copied like the other resources into `test-classes`. Probably, did I miss something?

Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: I'll try again.

Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: For this one, here is the code and the result in IntelliJ.
```
val path = s"src/test/resources/sqlgen/$answerFile.sql"
val path2 = getClass.getClassLoader.getResource("sqlgen/README.md")
```
```
file:/Users/dongjoon/SPARK-CHECK-GEN-SQL/sql/hive/target/scala-2.11/test-classes/sqlgen/README.md
```
Reading resource files are not problem, but we need the src directory for saving new golden SQL.

**Comment by dongjoon-hyun** [comment]
Now, the remaining issue is using `getResource` to save the golden files. I left a comment about that.

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala (hunk 6 above)
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: Ah ic. My worry is that Maven and SBT has base path where the tests are running, and the test will fail in Maven.

How about we define the output directory as a variable in the beginning of the suite, and then use that when we write output out? For reading input we simply use "sqlgen/$answerFile.sql"

Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: Also instead of using a variable as a flag, maybe we can use an environmental variable, and then document in the beginning

```
* To re-generate golden files, run:  SPARK_GENERATE_GOLDEN_FILES=1 build/sbt "hive/test-Only * LogicalPlanToSQLSuite"
```

something like this?

**Comment by rxin** [comment]
Can you also remove "[TEST]" from the title? TEST isn't a module.

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala (hunk 6 above)
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: Yep.

Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: While making that, I feel this becomes much convenient. Thank you.

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala [hunk 13]
```diff
@@ -17,15 +17,32 @@
...
+ * built-in functionality to automatically generate these golden files.
+ *
+ * To re-generate golden files, run:
+ *    SPARK_GENERATE_GOLDEN_FILES=1 build/sbt "hive/test-only *LogicalPlanToSQLSuite"
+ */
 class LogicalPlanToSQLSuite extends SQLBuilderTest with SQLTestUtils {
   import testImplicits._
 
+  // Used for generating new query answer files by saving
+  private val saveQuery = System.getenv("SPARK_GENERATE_GOLDEN_FILES") != null
```
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: hm if it is 0 we probably shouldn't be running it, so i'd test the value against 1 

```
Option(System.getenv("SPARK_GENERATE_GOLDEN_FILES")) == Some("1")

Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: also rename saveQuery to regenerateGoldenFiles

**Comment by dongjoon-hyun** [comment]
Thank you, @rxin . By the way, the following test occurs two times sequentially.
```
HiveSparkSubmitSuite.SPARK-8020: set sql conf in spark conf *** FAILED *** (5 minutes, 0 seconds)
```
I looked into the cases, but still have no idea about that. I wish the Jenkins pass at this time.

**Comment by dongjoon-hyun** [comment]
Thank you for guiding me in this PR, @rxin .

**Comment by liancheng** [comment]
LGTM.

One thing is that I feel most of the times the SQL comparison assertion may fail due to reasonable internal changes that somehow affect SQL generation in no harmful ways, and can be fixed by simply regenerating the golden files. That said, shall we add instructions about how to regenerate the golden files when the the assertion fails?

**Comment by dongjoon-hyun** [comment]
Thank you for review, @liancheng . Sure. Currently, it is only documented in class doc. I think you are suggesting to have that in some HTML or Wiki(Confluence). Did I understand your advice correctly?
```scala
/**
 * A test suite for LogicalPlan-to-SQL conversion.
 *
 * Each query has a golden generated SQL file in test/resources/sqlgen. The test suite also has
 * built-in functionality to automatically generate these golden files.
 *
 * To re-generate golden files, run:
 *    SPARK_GENERATE_GOLDEN_FILES=1 build/sbt "hive/test-only *LogicalPlanToSQLSuite"
 */
```

**Comment by dongjoon-hyun** [comment]
Hi, @rxin and @liancheng .
I will update this PR one more time. Please wait a moment.
I can use stable identifiers for `gen_attr`, too.

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/SQLBuilder.scala [hunk 14]
```diff
@@ -45,7 +46,12 @@ class SQLBuilder(logicalPlan: LogicalPlan) extends Logging {
   def this(df: Dataset[_]) = this(df.queryExecution.analyzed)
 
   private val nextSubqueryId = new AtomicLong(0)
+  private val nextGenAttrId = new AtomicLong(0)
   private def newSubqueryName(): String = s"gen_subquery_${nextSubqueryId.getAndIncrement()}"
+  private val exprIdMap: Map[Long, Long] = Map.empty[Long, Long]
+  private def normalizedName(n: NamedExpression): String = synchronized {
+    "gen_attr_" + exprIdMap.getOrElseUpdate(n.exprId.id, nextGenAttrId.getAndIncrement())
+  }
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/SQLBuilder.scala]: Now, the generated attributes also have stable identifiers like generated subqueries.

**Comment by dongjoon-hyun** [comment]
Hmm, HiveCompatibilitySuite has some dependency on `gen_attr_`. I reverted the last attempt.
I will try as another PR as planed before.

**Comment by dongjoon-hyun** [comment]
Jenkins is restarted, but the current last commit is efaa4d0 having the passed Jenkins test. Could you review and merge this first if possible, @rxin ?
```
Test build #62469 has finished for PR 14235 at commit efaa4d0.

This patch passes all tests.
This patch merges cleanly.
This patch adds no public classes.
```

**Comment by rxin** [comment]
Thanks - merging in master / 2.0.

I'm also merging this in 2.0 since it is a test only change and will reduce merge conflicts.

**Comment by dongjoon-hyun** [comment]
Oh, thank you for merging, @rxin ! Also, thank you for review, @gatorsmile and @liancheng .

**Comment by rxin** [comment]
@dongjoon-hyun can you also look into having stable identifiers for gen_attr? Right now the golden files look really weird because gen_attr is used more than once.

**Comment by dongjoon-hyun** [comment]
Sure. I've been looking that. It's on my list.
I'll make a JIRA issue and proceed.

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala [hunk 15]
```diff
@@ -17,15 +17,33 @@
...
+ *
+ * To re-generate golden files, run:
+ *    SPARK_GENERATE_GOLDEN_FILES=1 build/sbt "hive/test-only *LogicalPlanToSQLSuite"
+ */
 class LogicalPlanToSQLSuite extends SQLBuilderTest with SQLTestUtils {
   import testImplicits._
 
+  // Used for generating new query answer files by saving
+  private val regenerateGoldenFiles =
+    Option(System.getenv("SPARK_GENERATE_GOLDEN_FILES")).contains("1")
```
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: Why did you use contains here? This is super confusing and also broke 2.10.

I think I asked to do comparison with Some("1"). In most cases it is a very bad idea to use collection-oriented methods on Options, because they make the code more confusing.

Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: I fixed it here https://github.com/apache/spark/commit/c4524f5193e1b3ce1c56c5aed126f4121ce26d23

Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala]: Yep. I have nothing to say. My bad. Sorry about this. :(
